fake_spectra/spectra.py

```diff
@@ -673,28 +685,71 @@ def replace_not_DLA(self, ndla, thresh=10**20.3, elem="H", ion=1):
         cofm_DLA = np.empty_like(self.cofm)[:ndla, :]
         #Filter
         #Note: line does nothing
-        col_den = self.compute_spectra(elem, ion, 1215, False)
-        ind = self.filter_DLA(col_den, thresh)
+        col_den = self.compute_spectra(elem,ion,1215,False)
+        cdsum=np.sum(col_den, axis=1)
+
+        if self.MPI is not None:
+            # make sure array elements are contiguous
+            cdsum = np.ascontiguousarray(cdsum, np.float32)
+            # A variable for comm.Reduce()
+            cdsum_added = np.zeros_like(cdsum)
+            ### Call manager rank here
+            self.comm.Reduce(cdsum, cdsum_added, op=self.MPI.SUM, root=0)
+            size_ind = None
+            if self.rank == 0 :
+                ind = self.filter_DLA(cdsum_added, thresh)
+                size_ind = np.size(ind[0])
+            size_ind = self.comm.bcast(size_ind, root=0)
+
+            if self.rank != 0 :
+                ind = np.zeros(size_ind, dtype=np.int)
+            self.comm.Bcast(ind, root=0)
+        
+        else :
+            ind = self.filter_DLA(cdsum, thresh)
+
         H1_DLA = np.empty_like(col_den)
         #Update saves
         top = np.min([wanted, found+np.size(ind)])
         cofm_DLA[found:top] = self.cofm[ind][:top, :]
         H1_DLA[found:top] = col_den[ind][:top, :]
         found += np.size(ind)
         self.discarded = self.NumLos-np.size(ind)
+
         print("Discarded: ", self.discarded)
         while found < wanted:
             #Get a bunch of new spectra
             self.cofm = self.get_cofm()
-            col_den = self.compute_spectra(elem, ion, 1215, False)
-            ind = self.filter_DLA(col_den, thresh)
+            col_den = self.compute_spectra(elem,ion,1215,False)
+            cdsum = np.sum(col_den, axis = 1)
+            
+            if self.MPI is not None:
+                # make sure array elements are contiguous
+                cdsum = np.ascontiguousarray(cdsum, np.float32)
+                cdsum_added = np.zeros_like(cdsum)
+                ### Call manager rank here
+                self.comm.Reduce(cdsum, cdsum_added, op=self.MPI.SUM, root=0)
+                size_ind = None
+                if self.rank == 0 :
+                    ind = self.filter_DLA(cdsum_added, thresh)
+                    size_ind = np.size(ind[0])
+                size_ind = self.comm.bcast(size_ind, root=0)
```

Comment:
bcast vs Bcast?

Reply:
To broadcast a numpy array, I should know the size of the array on all ranks.  So, first I broadcast the size of the array with `bcast` to all ranks and then the the array itself with `Bcast`.

Reply:
I see: bcast is for python arrays, Bcast for numpy arrays and mpi4py doesn't work the size out for itself. Thanks.

Reply:
Exactly!